furi/core/event_loop.c

```diff
@@ -145,30 +218,225 @@ static FuriEventLoopProcessStatus
+
+    if(request->callback) {
+        instance->tick_interval = request->interval;
+        instance->tick_prev_time = xTaskGetTickCount();
+    }
+}
+
+static void furi_event_loop_process_request_queue(FuriEventLoop* instance) {
+    for(; !RequestQueue_empty_p(instance->request_queue);
+        RequestQueue_pop_back(NULL, instance->request_queue)) {
+        const FuriEventLoopRequestQueueItem* item = RequestQueue_back(instance->request_queue);
+
+        if(item->type == FuriEventLoopRequestTypeTimer) {
+            const FuriEventLoopTimerRequest* request = &item->timer_request;
+            furi_event_loop_process_timer_request(instance, request);
+
+        } else if(item->type == FuriEventLoopRequestTypeTick) {
+            const FuriEventLoopTickRequest* request = &item->tick_request;
+            furi_event_loop_process_tick_request(instance, request);
+
+        } else if(item->type == FuriEventLoopRequestTypePending) {
+            const FuriEventLoopPendingRequest* request = &item->pending_request;
+            request->callback(request->context);
+
+        } else {
+            furi_crash();
+        }
+    }
+}
+
+static void furi_event_loop_process_tick(FuriEventLoop* instance) {
+    if(instance->tick_callback && furi_event_loop_tick_is_expired(instance)) {
+        instance->tick_prev_time = xTaskGetTickCount();
+        instance->tick_callback(instance->tick_callback_context);
+    }
+}
+
+static void furi_event_loop_enqueue_request(
+    FuriEventLoop* instance,
+    const FuriEventLoopRequestQueueItem* item) {
+    RequestQueue_push_front(instance->request_queue, *item);
+
+    xTaskNotifyIndexed(
+        instance->thread_id, FURI_EVENT_LOOP_FLAG_NOTIFY_INDEX, FuriEventLoopFlagRequest, eSetBits);
+}
+
+static void furi_event_loop_restore_flags(FuriEventLoop* instance, uint32_t flags) {
+    if(flags) {
+        xTaskNotifyIndexed(
+            instance->thread_id, FURI_EVENT_LOOP_FLAG_NOTIFY_INDEX, flags, eSetBits);
+    }
+}
+
 void furi_event_loop_run(FuriEventLoop* instance) {
     furi_check(instance);
     furi_check(instance->thread_id == furi_thread_get_current_id());
 
-    uint32_t timeout = instance->tick_callback ? instance->tick_interval : FuriWaitForever;
-
     while(true) {
+        instance->state = FuriEventLoopStateIdle;
+
+        const TickType_t xTicksToWait = furi_event_loop_get_wait_time(instance);
+
         uint32_t flags = 0;
         BaseType_t ret = xTaskNotifyWaitIndexed(
-            FURI_EVENT_LOOP_FLAG_NOTIFY_INDEX, 0, FuriEventLoopFlagAll, &flags, timeout);
+            FURI_EVENT_LOOP_FLAG_NOTIFY_INDEX, 0, FuriEventLoopFlagAll, &flags, xTicksToWait);
 
         instance->state = FuriEventLoopStateProcessing;
+
         if(ret == pdTRUE) {
             if(flags & FuriEventLoopFlagStop) {
-                instance->state = FuriEventLoopStateIdle;
+                instance->state = FuriEventLoopStateStopped;
                 break;
+
             } else if(flags & FuriEventLoopFlagEvent) {
```

Comment:
else if -> if ?

Reply:
Makes no difference, since there is a break statement in the branch above.